sys/include/usb/hid.h

```diff
@@ -103,6 +105,375 @@ typedef struct __attribute__((packed)){
+
+/**
+ * @brief Application collection type
+ */
+#define USB_HID_COLLECTION_APPLICATION  0x01
+
+/**
+ * @brief Defines a HID descriptor short item.
+ *
+ * @param tag  Item tag
+ * @param type Item type
+ * @param size Item data size
+ */
+#define USB_HID_ITEM(tag, type, size) \
+    (((tag & 0xF) << 4) | ((type & 0x3) << 2) | (size & 0x3))
+
+/**
+ * @brief Defines a HID descriptor input item.
+ *
+ * @note data length is one byte
+ *
+ * @param input Input item flags
+ */
+#define USB_HID_INPUT(input)  \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_INPUT, USB_HID_ITEM_TYPE_MAIN, 1), \
+    input
+
+/**
+ * @brief Defines a HID descriptor output item.
+ *
+ * @note data length is one byte
+ *
+ * @param output Output item flags
+ */
+#define USB_HID_OUTPUT(output) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_OUTPUT, USB_HID_ITEM_TYPE_MAIN, 1), \
+    output
+
+
+/**
+ * @brief Defines a HID Feature item
+ *
+ * @note data length is one byte.
+ *
+ * @param feature Feature item data
+ */
+#define USB_HID_FEATURE(feature) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_FEATURE, USB_HID_ITEM_TYPE_MAIN, 1), \
+    feature
+
+/**
+ * @brief Defines a HID Collection item
+ *
+ * @note data length is one byte.
+ *
+ * @param collection Collection item data
+ */
+#define USB_HID_COLLECTION(collection) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_COLLECTION, USB_HID_ITEM_TYPE_MAIN, 1), \
+    collection
+
+/**
+ * @brief Defines a HID End Collection item
+ */
+#define USB_HID_END_COLLECTION \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_COLLECTION_END, USB_HID_ITEM_TYPE_MAIN, 0)
+
+/**
+ * @brief Defines a HID Usage Page item.
+ *
+ * @note data length is one byte.
+ *
+ * @param page Usage Page
+ */
+#define USB_HID_USAGE_PAGE(page) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_USAGE_PAGE, USB_HID_ITEM_TYPE_GLOBAL, 1), \
+    page
+
+/**
+ * @brief Defines a HID Usage Page item.
+ *
+ * @note data length is two bytes.
+ *
+ * @param page      Usage page
+ */
+#define USB_HID_USAGE_PAGE16(page) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_USAGE_PAGE, USB_HID_ITEM_TYPE_GLOBAL, 2), \
+    page & 0xff, (page >> 8)
+
+/**
+ * @brief Defines a HID Logical Minimum item
+ *
+ * @note data length is one byte.
+ *
+ * @param  minimum  Minimum value in logical units
+ */
+#define USB_HID_LOGICAL_MIN8(minimum) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_LOGICAL_MIN, USB_HID_ITEM_TYPE_GLOBAL, 1), \
+    minimum
+
+/**
+ * @brief Defines a HID Logical Minimum item
+ *
+ * @note data length is two bytes.
+ *
+ * @param  minimum  Minimum value in logical units
+ */
+#define USB_HID_LOGICAL_MIN16(minimum) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_LOGICAL_MIN, USB_HID_ITEM_TYPE_GLOBAL, 2), \
+    minimum & 0xff, (minimum >> 8)
+
+
+/**
+ * @brief Defines a HID Logical Minimum item
+ *
+ * @note data length is four bytes.
+ *
+ * @param  minimum  Minimum value in logical units
+ */
+#define USB_HID_LOGICAL_MIN32(minimum) \
+    USB_HID_ITEM(USB_HID_ITEM_TAG_LOGICAL_MIN, USB_HID_ITEM_TYPE_GLOBAL, 3), \
```

Comment:
Why 3 ? Shouldn't be 4 ??

Reply:
Section 6.2.2.2 defines that a data of 4 bytes long should be indicated with a size tag with value '3':

![image](https://user-images.githubusercontent.com/5160052/142684277-cd947429-adc6-4b93-a1ac-22778a6999bd.png)